velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/PrefixSort.h"
+
+using namespace facebook::velox::exec::prefixsort;
+
+namespace facebook::velox::exec {
+
+namespace {
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize() {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize<TypeKind::BIGINT>() {
+  return sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+}
+
+template <>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize<TypeKind::INTEGER>() {
+  return sizeof(TypeTraits<TypeKind::INTEGER>::NativeType);
+}
+
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize(const TypeKind& typeKind) {
+  return VELOX_DYNAMIC_TYPE_DISPATCH(normalizeSize, typeKind);
+}
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* const row,
+    char* const prefix) {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <TypeKind typeKind, typename nativeType>
+FOLLY_ALWAYS_INLINE void normalizeNumericColumn(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* const row,
+    char* const prefix) {
+  // Store null as min/max value according compare flags.
+  if (RowContainer::isNullAt(row, rowColumn.nullByte(), rowColumn.nullMask())) {
+    CompareFlags compareFlags = prefixSortLayout.compareFlags[index];
+    PrefixSortEncoder::encode<nativeType>(
```

Comment:
let's move null encoding logic into PrefixSortEncoder as well.

PrefixSortEncoder::encodeNull<nativeType>(layout.compareFlags, prefix + layout.prefixOffsets[index]);

Is it valid to treat null the same as min/max value? If we have 2 rows - null and min - how would we make sure the order comes out differently for nulls first and nulls last?

Reply:
> If we have 2 rows - null and min - how would we make sure the order comes out differently for nulls first and nulls last
This need a byte to handle null_first/last,  I do not consider this case before, and review the DuckDB implements:
```
{
	key_locations[i][0] = invalid;
	memset(key_locations[i] + 1, '\0', sizeof(T));
}
```

> PrefixSortEncoder::encodeNull(layout.compareFlags, prefix + layout.prefixOffsets[index]);

Maybe , need to add two method : 

PrefixSortEncoder::encodeNull(layout.compareFlags, prefix + layout.prefixOffsets[index]);
PrefixSortEncoder::encodeWithNull(layout.compareFlags, isNull,  prefix + layout.prefixOffsets[index]);

or combine them into encode method
PrefixSortEncoder::encode(layout.compareFlags,  hasNull,  isNull, prefix + layout.prefixOffsets[index]);

and how about seperate a Normalizer class to handle this logic, as duckDB scalter

![image](https://github.com/facebookincubator/velox/assets/13514695/55381274-06c9-40a8-89fc-fec7edd6d866)

Encoder still handle value encode logic ?

https://github.com/duckdb/duckdb/blob/main/src/common/row_operations/row_radix_scatter.cpp

@mbasmanova waiting for your response ~

Reply:
I suggest to modify PrefixSortEncoder::encode method to take std::optional<T>. std::nullopt would then indicate a null value. 

```
void encode(std::optional<T> value, char* row)
```

I do not think it is worth trying to save a null byte at this point. If we do need to do that, then we can have an extra method:

```
void encodeNoNulls(T value, char* row)
```

Reply:
ok, I will open a new PR first to modify the method~